pyro/infer/autoguide/guides.py

```diff
@@ -583,6 +583,133 @@ def _loc_scale(self, *args, **kwargs):
         return loc, scale
 
 
+class AutoProgressive(AutoContinuous):
+    """
+    This implementation of :class:`AutoContinuous` progressively increases
+    guide complexity through three stages of training:
+
+    1.  This guide starts learning a point estimate, equivalent to
+        :class:`AutoDelta` .
+    2.  Next, after :meth:`init_scale` is called, this additionally learns
+        a diagonal noise scale, equivalent to :class:`AutoDiagonalNormal` .
+        During this stage the location parameter continues to be learned.
+    3.  Finally, after :meth:`init_cov` is called, this additionally learns a
+        low-rank covariance structure among latent variables, equivalent to
+        :class:`AutoLowRankMultivariateNormal` . During this stage the location
+        and diagonal noise scale parameters continue to be learned.
+
+    Usage::
+
+        # Begin stage 1.
+        guide = AutoProgressive(model, rank=10)
+        svi = SVI(model, guide, ...)
+        for _ in range(100):
+            svi.step(data)
+
+        # Move to stage 2.
+        guide.init_scale()
+        for _ in range(100):
+            svi.step(data)
+
+        # Move to stage 3.
+        guide.init_cov()
+        for _ in range(100):
+            svi.step(data)
+
+    The guide does not depend on the model's ``*args, **kwargs``.
+
+    :param callable model: A generative model.
+    :param int rank: The rank of the low-rank part of the covariance matrix.
+    :param str prefix: A prefix for all internal param sites.
+    :param callable init_loc_fn: A per-site initialization function.
+        See :ref:`autoguide-initialization` section for available functions.
+    """
+
+    def __init__(self, model, prefix="auto", init_loc_fn=init_to_median, rank=1):
+        if not isinstance(rank, int) or not rank > 0:
+            raise ValueError("Expected rank > 0 but got {}".format(rank))
```

Comment:
what if someone just wants to do stages 1 + 2?

Reply:
Works fine. I've added a comment to that effect.